File: 290. Word Pattern/Solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        istringstream split(s);
        vector<string> words;
        string w;
        while(split>>w){words.push_back(w);}
        if(words.size() != pattern.size()){return false;}
        unordered_map<char,string> char_str;
        unordered_map<string,char> str_char;
        for(int i=0; i<pattern.size(); i++){
            char chr = pattern[i];
            string sr = words[i];
            if(char_str.count(chr)){if(char_str[chr] != sr){return false;}}
            else{char_str[chr] = sr;}
            if(str_char.count(sr)){if(str_char[sr] != chr){return false;}}
            else{str_char[sr] = chr;}
        }
        return true;
    }
};

File: 290. Word Pattern/Solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words): return False
        char_str,str_char = {}, {}
        for i in range(len(pattern)):
            chr,sr = pattern[i],words[i]
            if chr in char_str:
                if char_str[chr] != sr: return False
            else: char_str[chr] = sr
            if sr in str_char:
                if str_char[sr] != chr: return False
            else: str_char[sr] = chr
        return True
